### PR TITLE
Flesh out framework for RAII-able ICU objects

### PR DESCRIPTION
**Pull request by jackhorton**

### CODE REVIEW

**Comment by jackhorton** [comment]
I can make this a code comment if you want but I will just put it here for now.

In general there is only one new class, `ScopedICUObject<T, Function>`, which is only meant to be used in one of its typedef'd forms, which all take the naming form of `Scoped%ICUObjectName%`, like `ScopedUCollator`. Its fairly straightforward, as its constructor takes a single argument of type T and has an `operator T` such that the only update it requires compared to not using any wrapper is changing `ICUObject a = open_icuobject()` to `ScopedICUObject a(open_icuobject())`. There are some template/macro/inline shenanigans because C++ is awful, but its API should be mostly invisible.

FinalizableICUObject used to be used directly and required the relevant close function to be passed. I updated it to use the same template/macro/inline shenanigans that ScopedICUObject uses, such that it is now designed to be used in one of the typedef'd versions like FinalizableUNumberFormat

**Comment by jackhorton** [comment]
And I went with the template/macro/inline + typedef shenanigans because I assumed that there was more runtime overhead of duplicating the classes, but I suppose I didn't test anything ahead of time to confirm that assumption.

File: lib/Runtime/PlatformAgnostic/ICU.h [hunk 1]
```diff
@@ -22,65 +22,47 @@ namespace PlatformAgnostic
...
+        };
+    }
+}
 
-#define INSTANTIATE_AUTOICUOBJECT(T, Closer) \
+#define INSTANTIATE_SCOPEDICUOBJECT(TObject, CloseFunction) \
         template<> \
-        constexpr AutoICUObject<T *>::CloseFunction CloseFunctionForType(T *) \
+        inline PlatformAgnostic::ICUHelpers::ScopedICUObject<TObject *, CloseFunction>::ScopedICUObject(TObject * object) : \
+            object(object) \
```
Comment (sethbrenith) [inline lib/Runtime/PlatformAgnostic/ICU.h]: Any particular reason these constructors need to be defined in macros rather than in the main class template?

Comment (jackhorton) [reply lib/Runtime/PlatformAgnostic/ICU.h]: Just realized that they don't anymore -- the template parameter is enough, which makes this all actually much cleaner. Thanks!

**Pull request merged**